agents/ModularAgent.py

```diff
@@ -11,434 +11,509 @@
+                                                 state_featurized)
+
+        if(len(explanations) == 0):
+
+            if(len(nonmatching_explanations) > 0):
+                explanations = [choice(nonmatching_explanations)]
+
+            else:
+                explanations = self.explanations_from_how_search(
+                               state_featurized, sai, foci_of_attention)
+
+                explanations = self.which_learner.cull_how(explanations)
+
+                rhs_by_how = self.rhs_by_how.get(skill_label, {})
+                for exp in explanations:
+                    if(exp.rhs.as_tuple in rhs_by_how):
+                        exp.rhs = rhs_by_how[exp.rhs.as_tuple]
+                    else:
+                        rhs_by_how[exp.rhs.as_tuple] = exp.rhs
+                        self.rhs_by_how[skill_label] = rhs_by_how
+                        self.add_rhs(exp.rhs)
+
+        self.fit(explanations, state_featurized, reward)
+
+    # ------------------------------CHECK--------------------------------------
+
+    def check(self, state, sai):
+        state_featurized, knowledge_base = self.planner.apply_featureset(state)
+        explanations = self.explanations_from_skills(state, sai, self.rhs_list)
+        explanations, _ = self.where_matches(explanations)
+        return len(explanations) > 0
+
+    def get_skills(self, states=None):
+        out = []
+        for state in states:
+            req = self.request(state,
+                               add_skill_info=True).get('skill_info', None)
+
+            if(req is not None):
+                out.append(frozenset([(k, v) for k, v in req.items()]))
+
+        uniq_lst = list(dict.fromkeys(out).keys())
+        unique = [{k: v for k, v in x} for x in uniq_lst]  # set(out)]
+        return unique
+
+
+# ---------------------------CLASS DEFINITIONS---------------------------------
+
+def ground(arg):
+    """
+    Doc String
+    """
+    if isinstance(arg, tuple):
+        return tuple(ground(e) for e in arg)
+    elif isinstance(arg, str):
+        return arg.replace('?', 'QM')
+    else:
+        return arg
+
+
+def unground(arg):
+    """
+    Doc String
+    """
+    if isinstance(arg, tuple):
+        return tuple(unground(e) for e in arg)
+    elif isinstance(arg, str):
+        return arg.replace('QM', '?')
+    else:
+        return arg
+
+
+def flatten_state(state):
+    tup = Tuplizer()
+    flt = Flattener()
+    state = flt.transform(tup.transform(state))
+    return state
+
+
+def grounded_key_vals_state(state):
+    return [(ground(a), state[a].replace('?', 'QM')
+            if isinstance(state[a], str)
+            else state[a])
+            for a in state]
+
+
+def kb_to_flat_ungrounded(knowledge_base):
+    state = {unground(a): v.replace("QM", "?")
+             if isinstance(v, str)
+             else v
+             for a, v in knowledge_base.facts}
+    return state
+
+
+class StateMultiView(object):
```

Comment:
We can skip it for now but some time I would like to look at extracting some of these helper classes to their own files so they can be used more broadly.

Reply:
#36